File: tmp_parse_robinhood.py
import csv
import re
from datetime import datetime

path = '/Users/mcraig/Downloads/robinhood/Individual-a64d70b9-5c97-5ef7-a021-3bf8fed56c8d.csv'

def parse_date(s):
    s = s.strip()
    if not s:
        return None
    for fmt in ('%m/%d/%Y','%m/%d/%y','%Y-%m-%d'):
        try:
            return datetime.strptime(s, fmt).date()
        except Exception:
            pass
    try:
        # try generic
        return datetime.fromisoformat(s).date()
    except Exception:
        return None


def normalize_desc(s):
    return re.sub(r"\s+", " ", (s or '').strip().lower())


def parse_instrument(instr, desc):
    s = (instr or '').strip()
    if not s:
        s = desc or ''
    s = re.sub(r"\s+", " ", s)
    ticker = ''
    strike = None
    type_ = None
    expiry = None

    # Prefer $NN(.NN) for strike
    m = re.search(r"\$\s*([0-9]+(?:\.[0-9]+)?)", s)
    if m:
        strike = float(m.group(1))
    else:
        # fallback: number immediately before Put/Call
        m2 = re.search(r"(\d+(?:\.\d+)?)\s+(?=Put|Call|P\b|C\b)", s, re.I)
        if m2:
            strike = float(m2.group(1))

    # Expiry detection
    exp_m = re.search(r"(\d{1,2}/\d{1,2}/\d{2,4})", s) or re.search(r"(\d{4}-\d{2}-\d{2})", s)
    if exp_m:
        d = parse_date(exp_m.group(1))
        if d:
            expiry = d.isoformat()

    # Type
    type_m = re.search(r"\b(put|call|p|c)\b", s, re.I)
    if type_m:
        type_ = 'Put' if re.search(r"put", type_m.group(1), re.I) or re.search(r"^p$", type_m.group(1), re.I) else 'Call'

    # Ticker: prefer leading token if alphabetic, else first alpha token that's not a common word
    parts = s.split(' ')
    if parts:
        if re.match(r'^([A-Za-z\.]{1,6})$', parts[0]):
            ticker = parts[0].upper()
    if not ticker:
        for t in parts:
            if re.match(r'^[A-Za-z\.]{1,6}$', t) and not re.search(r"\b(put|call|sold|buy|to|close|option|expiration|cusip|for)\b", t, re.I):
                ticker = t.upper()
                break

    return {'ticker': ticker or '', 'strike': strike, 'type': type_, 'expiry': expiry}


def derive_price(amount, quantity):
    # amount like 116.97 or (223.02), quantity int
    if not amount:
        return 0.0
    a = amount.replace('(', '').replace(')', '').replace('$','').replace(',','')
    try:
        amt = float(a)
    except Exception:
        return 0.0
    qty = abs(int(quantity)) if quantity and quantity.strip()!='' else 1
    price = abs(amt) / (qty * 100.0)
    return price

# Read CSV
with open(path, newline='') as f:
    reader = csv.DictReader(f)
    rows = list(reader)

print('Total rows:', len(rows))

# find keys
keys = [k for k in rows[0].keys()]
print('Headers:', keys)

# canonicalize
canonical = []
for r in rows:
    if not r:
        continue
    # Defensive: keys may be missing for some malformed rows
    canonical.append({
        'activityDate': (r.get('Activity Date') or '').strip(),
        'description': (r.get('Description') or '').strip(),
        'amount': (r.get('Amount') or '').strip(),
        'quantity': (r.get('Quantity') or '').strip(),
        'instrument': (r.get('Instrument') or '').strip(),
        'trans': (r.get('Trans Code') or '').strip(),
        'process_date': (r.get('Process Date') or '').strip(),
        'settle_date': (r.get('Settle Date') or '').strip(),
        'raw': r
    })

# Filter out rows that are Process/Settle? The sample has Process Date/Settle Date columns but their presence doesn't mean row should be ignored.
# We'll ignore rows without a trans code or without description containing 'Put'/'Call' for options when looking for STO/BTC.

filtered = canonical
print('Filtered rows (non-empty):', len(filtered))

sto_rows = [r for r in filtered if (r['trans'].upper()=='STO') and re.search(r'put', r['description'], re.I)]
btc_rows = [r for r in filtered if r['trans'].upper()=='BTC']

print('STO rows:', len(sto_rows), 'BTC rows:', len(btc_rows))

trades = []
for sto in sto_rows:
    parsed = parse_instrument(sto['description'] or sto['instrument'], sto['description'])
    qty = abs(int(sto['quantity'])) if sto['quantity'] else 1
    entry = derive_price(sto['amount'], sto['quantity'])
    ticker = parsed['ticker'] or (sto['instrument'] or '').split(' ')[0]
    strike = parsed['strike']
    expiry = parsed['expiry']
    openedDate = sto['activityDate']
    trade = {
        'ticker': (ticker or '').upper(),
        'type':'CSP',
        'strike': strike,
        'quantity': qty,
        'entryPrice': entry,
        'closePrice': 0.0,
        'openedDate': openedDate,
        'expirationDate': expiry,
        'closedDate': None,
        'status': 'Open',
        '_desc': sto['description']
    }
    # detect strict roll
    norm = normalize_desc(sto['description'])
    sto_date = parse_date(sto['activityDate'])
    candidate = None
    for btc in btc_rows:
        if normalize_desc(btc['description']) != norm:
            continue
        btc_date = parse_date(btc['activityDate'])
        if not btc_date or not sto_date:
            continue
        if btc_date < sto_date:
            continue
        btc_parsed = parse_instrument(btc['instrument'] or btc['description'], btc['description'])
        if strike and btc_parsed['strike'] and abs(float(strike) - float(btc_parsed['strike'])) > 1e-6:
            continue
        if expiry and btc_parsed['expiry'] and expiry != btc_parsed['expiry']:
            continue
        candidate = btc
        break
    if candidate:
        trade['status'] = 'Rolled'
        trade['closedDate'] = candidate['activityDate']
        trade['closePrice'] = derive_price(candidate['amount'], candidate['quantity'])
    trades.append(trade)

print('Derived trades:', len(trades))
print('Sample trades (first 10):')
for t in trades[:10]:
    print(t)

rolled = [t for t in trades if t['status']=='Rolled']
print('Rolled count:', len(rolled))
if rolled:
    print('Rolled examples:')
    for r in rolled[:5]:
        print(r)

missing_ticker = [t for t in trades if not t['ticker']]
print('Trades missing ticker:', len(missing_ticker))
if missing_ticker:
    print('Examples missing ticker:', missing_ticker[:3])

print('Done')

File: tmp_parse_robinhood.js
const fs = require('fs');
const path = '/Users/mcraig/Downloads/robinhood/Individual-a64d70b9-5c97-5ef7-a021-3bf8fed56c8d.csv';
const text = fs.readFileSync(path, 'utf8');

// parseCSV from App.jsx
const parseCSV = (text) => {
    const rows = [];
    let cur = '';
    let row = [];
    let inQuotes = false;
    for (let i = 0; i < text.length; i++) {
        const ch = text[i];
        const next = text[i + 1];
        if (ch === '"') {
            if (inQuotes && next === '"') { // escaped quote
                cur += '"';
                i++;
            } else {
                inQuotes = !inQuotes;
            }
        } else if (ch === ',' && !inQuotes) {
            row.push(cur);
            cur = '';
        } else if ((ch === '\n' || ch === '\r') && !inQuotes) {
            if (cur !== '' || row.length > 0) {
                row.push(cur);
                rows.push(row);
                row = [];
                cur = '';
            }
            if (ch === '\r' && next === '\n') i++;
        } else {
            cur += ch;
        }
    }
    if (cur !== '' || row.length > 0) {
        row.push(cur);
        rows.push(row);
    }

    if (rows.length === 0) return [];
    const headers = rows[0].map(h => h.trim());
    return rows.slice(1).map(r => {
        const obj = {};
        for (let i = 0; i < headers.length; i++) {
            obj[headers[i]] = (r[i] || '').trim();
        }
        return obj;
    });
};

const normalizeDesc = (s) => (s || '').toLowerCase().replace(/\s+/g, ' ').trim();

const parseInstrument = (instrument) => {
    if (!instrument) return {};
    const res = { ticker: '', strike: null, type: '', expiry: '' };
    const parts = instrument.replace(/\s+/g, ' ').trim().split(' ');
    if (parts.length > 0 && /^[A-Za-z\.]{1,6}$/.test(parts[0])) {
        res.ticker = parts[0].toUpperCase();
    }
    const strikePart = parts.find(p => /^\d+(?:\.\d+)?$/.test(p.replace(/[^0-9.]/g, '')));
    if (strikePart) res.strike = Number(strikePart.replace(/[^0-9.]/g, ''));
    const typePart = parts.find(p => /put|call|p|c/i.test(p));
    if (typePart) res.type = /put/i.test(typePart) || /^p$/i.test(typePart) ? 'Put' : 'Call';
    const expiryPart = parts.find(p => /\d{1,2}[/\-]\d{1,2}[/\-]\d{2,4}/.test(p) || /\d{4}-\d{2}-\d{2}/.test(p));
    if (expiryPart) {
        const m = expiryPart.match(/(\d{1,2})[/\-](\d{1,2})[/\-](\d{2,4})/);
        if (m) {
            let [_, mm, dd, yy] = m;
            if (yy.length === 2) yy = '20' + yy;
            res.expiry = `${yy.padStart(4, '0')}-${mm.padStart(2, '0')}-${dd.padStart(2, '0')}`;
        } else {
            const n = expiryPart.match(/(\d{4})-(\d{2})-(\d{2})/);
            if (n) res.expiry = `${n[1]}-${n[2]}-${n[3]}`;
        }
    }
    return res;
};

const parseRobinhoodRows = (rows) => {
    const sample = rows[0] || {};
    const keys = Object.keys(sample);
    const k = (nameCandidates) => keys.find(h => nameCandidates.some(n => h.toLowerCase().includes(n)));

    const dateKey = k(['activity date', 'activitydate', 'date']);
    const descKey = k(['description', 'desc']);
    const amountKey = k(['amount', 'net amount', 'price']);
    const qtyKey = k(['quantity', 'qty']);
    const instrKey = k(['instrument', 'symbol', 'instrument symbol']);
    const transKey = k(['trans code', 'transcode', 'transaction', 'activity type', 'action']);
    const processKey = k(['process', 'settle', 'process/settle']);

    const canonicalRows = rows.map(r => ({
        activityDate: r[dateKey] || r[dateKey?.toLowerCase()] || '',
        description: r[descKey] || '',
        amount: (r[amountKey] || '').replace(/[^0-9\.-]/g, ''),
        quantity: (r[qtyKey] || '').replace(/[^0-9\.-]/g, ''),
        instrument: r[instrKey] || '',
        trans: r[transKey] || '',
        process: r[processKey] || '',
        raw: r,
    }));

    const filtered = canonicalRows.filter(rr => {
        const p = (rr.process || '').toLowerCase();
        return !(p.includes('process') || p.includes('settle'));
    });

    const stoRows = filtered.filter(rr => (rr.trans || '').toUpperCase() === 'STO' && /put/i.test(rr.description));
    const btcRows = filtered.filter(rr => (rr.trans || '').toUpperCase() === 'BTC');

    const trades = [];

    const derivePrice = (row) => {
        const amt = Number(row.amount) || 0;
        const qty = Math.abs(Number(row.quantity)) || 1;
        const price = Math.abs(amt) / (qty * 100) || 0;
        return price;
    };

    stoRows.forEach(sto => {
        const parsed = parseInstrument(sto.instrument || sto.description);
        const qty = Math.abs(Number(sto.quantity)) || 1;
        const entry = derivePrice(sto);
        const ticker = parsed.ticker || (sto.instrument || '').split(' ')[0] || '';
        const strike = parsed.strike || null;
        const expiry = parsed.expiry || '';
        const openedDate = sto.activityDate;

        const trade = {
            ticker: ticker.toUpperCase(),
            type: 'CSP',
            strike: strike,
            quantity: qty,
            entryPrice: entry,
            closePrice: 0,
            openedDate,
            expirationDate: expiry,
            closedDate: null,
            status: 'Open',
            _desc: sto.description,
        };

        const norm = normalizeDesc(sto.description);
        const candidate = btcRows.find(btc => {
            if (normalizeDesc(btc.description) !== norm) return false;
            const stoDate = new Date(openedDate);
            const btcDate = new Date(btc.activityDate);
            if (isNaN(stoDate) || isNaN(btcDate)) return false;
            if (btcDate < stoDate) return false;
            const btcParsed = parseInstrument(btc.instrument || btc.description);
            if (strike && btcParsed.strike && Number(strike) !== Number(btcParsed.strike)) return false;
            if (expiry && btcParsed.expiry && expiry !== btcParsed.expiry) return false;
            return true;
        });

        if (candidate) {
            trade.status = 'Rolled';
            trade.closedDate = candidate.activityDate;
            trade.closePrice = derivePrice(candidate);
        }

        trades.push(trade);
    });

    return { trades, counts: { sto: stoRows.length, btc: btcRows.length, total: filtered.length } };
};

const rows = parseCSV(text);
console.log('Parsed rows:', rows.length);
const { trades, counts } = parseRobinhoodRows(rows);
console.log('Filtered rows (non-process/settle):', counts.total, 'STO:', counts.sto, 'BTC:', counts.btc);
console.log('Trades derived:', trades.length);
console.log('First 10 trades (JSON):\n', JSON.stringify(trades.slice(0, 10), null, 2));

// Show rolled trades
const rolled = trades.filter(t => t.status === 'Rolled');
console.log('Rolled trades count:', rolled.length);
if (rolled.length) console.log('Rolled sample:\n', JSON.stringify(rolled.slice(0,3), null, 2));

// Check for parsing issues
const missingTicker = trades.filter(t => !t.ticker);
console.log('Trades missing ticker:', missingTicker.length);
if (missingTicker.length) console.log(JSON.stringify(missingTicker.slice(0,3), null, 2));

console.log('Done.');
